refactor: Improve design of task rendering code & add tests

esm7 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

# Description

As part of https://github.com/obsidian-tasks-group/obsidian-tasks/discussions/877, this PR offers a pure refactor (no change in functionality) to the code of task rendering to HTML, and tests coverage to areas that were previously unreachable by the tests.

This includes:

- A new `TaskLineRenderer.ts` file that encapsulates the rendering flow of a task, from a `Task` object into an HTML. Some code from `Task` moved to here.
- Expansion of the existing `LayoutOptions` concept into a `TaskLayout` class. The legacy `LayoutOptions` class translates directly to query options, but it does not represent the complete product (it says what to display or hide, but not how the task should eventually look like). `LayoutOptions` was moved into `TaskLayout.ts` and a `TaskLayout` can be created by a `LayoutOptions` object.
- The conversion of a task to string (`Task.toString`) now passes through a `TaskLayout`, which formalizes the structure of the rendered task, and renders each component separately with a `componentToString` method. This enables to reuse the same code when other layout combinations are required.
- Tests that utilize this new structure to cover the rendering flow of a task and its layout options.

## Motivation and Context

As agreed in the discussion above, I split the work on adding theming capabilities to Tasks into a pure refactor PR (this one) and a 2nd PR that will build on top of these changes to add fine-grained classes to task components.

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

Tests were added to the rendering flow.

## Screenshots (if appropriate)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

Changes visible to users:

- [ ] **Bug fix** (prefix: `fix` - non-breaking change which fixes an issue)
- [ ] **New feature** (prefix: `feat` - non-breaking change which adds functionality)
- [ ] **Breaking change** (prefix: `feat!!` or `fix!!` - fix or feature that would cause existing functionality to not work as expected)
- [ ] **Documentation** (prefix: `docs` - improvements to any documentation content)
- [ ] **Sample vault** (prefix: `vault` - improvements to the [Tasks-Demo sample vault](https://github.com/obsidian-tasks-group/obsidian-tasks/tree/main/resources/sample_vaults/Tasks-Demo))

Internal changes:

- [x] **Refactor** (prefix: `refactor` - non-breaking change which only improves the design or structure of existing code, and making no changes to its external behaviour)
- [x] **Tests** (prefix: `test` - additions and improvements to unit tests and the smoke tests)
- [ ] **Infrastructure** (prefix: `chore` - examples include GitHub Actions, issue templates)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project and passes `yarn run lint`.
- [ ] My change requires a change to the documentation.
- [ ] I have [updated the documentation](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CONTRIBUTING.md#updating-documentation) accordingly.
- [x] My change has adequate [Unit Test coverage](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CONTRIBUTING.md#maintaining-the-tests).

## Terms

<!--
By submitting this pull request, you must agree to follow our
[contributing guide](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CONTRIBUTING.md) and
[Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md).
Put an x in the boxes to confirm you agree.
-->

- [x] My contribution follow this project's [contributing guide](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CONTRIBUTING.md)
- [x] I agree to follow this project's [Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md)

esm7 commented:
I would appreciate anyone who has the time to review and test this branch. I know I need to update it (it's from a few weeks back), but other than that, queries here should work exactly the same as before this refactor. Boring but important ;)

claremacrae commented:
Wow, brilliant, thanks @esm7.

I will of course review it. I should be able to get to it in the next day or two.

Do you have any suggestions as to how I can efficiently review the code that changed in files that got renamed please?
The first one I noticed is `LayoutOptions` to `TaskLayout`?
I normally review the diffs side-by-side in GitHub and my IDE but that won't work in this case?

claremacrae commented:
> Do you have any suggestions as to how I can efficiently review the code that changed in files that got renamed please? The first one I noticed is `LayoutOptions` to `TaskLayout`? I normally review the diffs side-by-side in GitHub and my IDE but that won't work in this case?

Ooohhh - maybe on my machine, with this branch checked out, I could use WebStorm refactoring to rename the new name (file, class, whatever) back to the old one temporarily... Then I can diff against the code at the start of your branch and it should be easy to see what else changed other than the file names...

I think that should work.

claremacrae commented:
Closing and reopening to try and force the GitHub Actions to run.

claremacrae commented:
Odd. I wonder if the tests won't run until the conflicts have been resolved?

claremacrae commented:
> I will of course review it. I should be able to get to it in the next day or two.

Just to set expectations... I don't have much time available over the next few days.
And I always review PRs thoroughly.
For a PR as large and wide-ranging as this, I see I'm gonna need more than a day or two of free time.
But I will definitely make it a priority to review it!

claremacrae commented:
> ## How has this been tested?
> Tests were added to the rendering flow.

Could you give me an idea of how much testing you've done to confirm that the appearance is indeed unchanged please? 
For example, what kinds of combinations of layout options have you tested with?
Thanks.

claremacrae commented on src/Task.ts:
```diff
@@ -438,159 +439,78 @@ export class Task {
+    /**
+     * Renders a specific LayoutComponent of the task (its description, priority, etc) as a string.
+     */
+    componentToString(layout: TaskLayout, component: LayoutComponent) {
```
Trivia: Please would you add `public` here?
Even though I don't always remember myseful, I'm trying to stick with the original policy of this project to include `public` even though it's the default.

claremacrae commented on src/TaskLayout.ts:
```diff
@@ -0,0 +1,72 @@
+    shortMode: boolean = false;
+}
+
+export type LayoutComponent =
+    | 'description'
+    | 'priority'
+    | 'recurrenceRule'
+    | 'startDate'
+    | 'scheduledDate'
+    | 'dueDate'
+    | 'doneDate'
+    | 'blockLink';
+
```
I was confused as to `whyLayoutComponent` did not include all the values in LayoutOptions.

From reading the code, I think the answer is probably that that it only contains
the options that reflect the contents of a rendered Task.

If that is correct, I think things would be clearer if this was named something like `TaskLayoutComponent`

esm7 replied:
Will do. You understood correctly and `TaskLayoutComponent` is indeed better.

claremacrae commented on src/TaskLayout.ts:
```diff
@@ -0,0 +1,72 @@
+/**
+ * This represents the desired layout of tasks when they are rendered in a given configuration.
+ * The layout is used when flattening the task to a string and when rendering queries, and can be
+ * modified by applying LayoutOptions objects.
```
```suggestion
  * modified by applying {@link LayoutOptions} objects.
```

claremacrae replied:
It helps navigating in WebStorm if symbol names in JSdoc are @linked... (though the @link is not needed in @see lines).

esm7 replied:
I can try to add them manually, just note that I don't have these features to test in my own IDE (which is `nvim`), so it's a hit or miss :shrug:

claremacrae replied:
You can always just commit my suggested edit above, I think - perhaps after merging and pushing my branch.

But it's really easy for me to add them, so no worries at all if you don't link them.

claremacrae commented on src/TaskLayout.ts:
```diff
@@ -0,0 +1,72 @@
+    public options: LayoutOptions;
+
+    constructor(options?: LayoutOptions, components?: LayoutComponent[]) {
+        if (options) this.options = options;
+        else this.options = new LayoutOptions();
+        if (components) this.layoutComponents = components;
+        else this.layoutComponents = this.defaultLayout;
```
I value easy readability of code greatly, and almost all uses of one-line ifs without {}, I find that the structure becomes less clear.

In this particular case, I spent some time looking at these 4 lines and trying to work out what the pattern was in the code.

Eventually I added {} and the hidden pattern immediately became clear, because the `this.options` is aligned. Such patterns really help.

```ts
       if (options) {
            this.options = options;
        } else {
            this.options = new LayoutOptions();
        }
        if (components) {
            this.layoutComponents = components;
        } else {
            this.layoutComponents = this.defaultLayout;
        }
```

Having seen the above, I realised that another option might be to use each of the 2 blocks a one-liner by using a ternary operator.

claremacrae commented on src/TaskLayout.ts:
```diff
@@ -0,0 +1,72 @@
+            return array.filter((element) => element != value);
+        };
+        let newComponents = this.layoutComponents;
+        if (layoutOptions.hidePriority) newComponents = arrayRemove(newComponents, 'priority');
+        if (layoutOptions.hideRecurrenceRule) newComponents = arrayRemove(newComponents, 'recurrenceRule');
+        if (layoutOptions.hideStartDate) newComponents = arrayRemove(newComponents, 'startDate');
+        if (layoutOptions.hideScheduledDate) newComponents = arrayRemove(newComponents, 'scheduledDate');
+        if (layoutOptions.hideDueDate) newComponents = arrayRemove(newComponents, 'dueDate');
+        if (layoutOptions.hideDoneDate) newComponents = arrayRemove(newComponents, 'doneDate');
+        return newComponents;
```
The pattern in names here is hard to see - for example to glance and check that the names are consistent on each line. I think it is because of the jagged/inconsistent alignment in the if condition.

I tried putting in {} and got this - which allows glancing down the `layout.hideBlah` lines and the option names very quickly.

```ts
        if (layoutOptions.hidePriority) {
            newComponents = arrayRemove(newComponents, 'priority');
        }
        if (layoutOptions.hideRecurrenceRule) {
            newComponents = arrayRemove(newComponents, 'recurrenceRule');
        }
        if (layoutOptions.hideStartDate) {
            newComponents = arrayRemove(newComponents, 'startDate');
        }
        if (layoutOptions.hideScheduledDate) {
            newComponents = arrayRemove(newComponents, 'scheduledDate');
        }
        if (layoutOptions.hideDueDate) {
            newComponents = arrayRemove(newComponents, 'dueDate');
        }
        if (layoutOptions.hideDoneDate) {
            newComponents = arrayRemove(newComponents, 'doneDate');
        }
```

Another option would be to move the if condition inside the local function, so it looked something approximately like:

```ts
newComponents = arrayRemove(newComponents, layoutOptions.hidePriority, 'priority');
newComponents = arrayRemove(newComponents, layoutOptions.hideRecurrenceRule, 'recurrenceRule');
newComponents = arrayRemove(newComponents, layoutOptions.hideStartDate, 'startDate');
newComponents = arrayRemove(newComponents, layoutOptions.hideScheduledDate, 'scheduledDate');
newComponents = arrayRemove(newComponents, layoutOptions.hideDueDate, 'dueDate');
newComponents = arrayRemove(newComponents, layoutOptions.hideDoneDate, 'doneDate');
```

It's probably obvious but I intentionally put the `layoutOptions.hide...` after the newComponents argument so the alignment aided readability.

Now any inconsistent values - like mixing due and start date on a line due to copying and pasting an existing line when new options are added in future, will jump out easily.

esm7 replied:
Will fix.

claremacrae commented on src/TaskLineRenderer.ts:
```diff
@@ -0,0 +1,228 @@
+import { Component, MarkdownRenderer } from 'obsidian';
+import type { Moment } from 'moment';
+import type { Task } from './Task';
+import * as taskModule from './Task';
```
For my own education: What's the benefit of importing like this?
And is there a significance of `taskModule` begin with lower case?

esm7 replied:
I find `import * as` useful for preventing namespace pollution. If I import individual classes from `Task` and use them straight out, e.g. `Status` or `Priority`, it can sometimes lose context to where it belongs, and `taskModule.Status` may be clearer.
I don't know what's the naming convention here, it seemed natural to me that an imported module is equivalent to an object so it begins with a lowercase (unlike a class for example), which is consistent with the common old Node.JS `require` clause.

claremacrae replied:
Oh wow, I spend a lot of time navigating to the location of a few functions with common names - seeing them called via the module/file name would really help comprehension.

I would go so far as to say that I would like to evolve the whole code base to that style. 

Related: I spend a lot of time futzing around adding or removing functions from import statements and that would go too.

I'm arbitrarily going to suggest that module names are capitalised like classes - as they feel more like classes and less like variables to me.

claremacrae commented on src/TaskLineRenderer.ts:
```diff
@@ -0,0 +1,228 @@
+    renderDetails.parentUlElement.appendChild(li);
+    li.classList.add('task-list-item', 'plugin-tasks-list-item');
+
+    const textSpan = document.createElement('span');
```
In my initial notes I said:

> I don't know enough about rendering to know if this a change or not, but I am looking for things that might be changes in behaviour.
> I cannot find any reference to 'span' in the original code when I search for 6ad76001c2e0e9361219196c6809cf9094800081.
> Bit worried that this might break user CSS filters???
> Would anything need to be changed in https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/docs/advanced/styling.md ?

But when I ran the new version in Obsidian with the Developer Tools, I couldn't see any difference in what was generated...

So I think it's all OK...

esm7 replied:
This is completely equivalent to the `const textSpan = li.createSpan();` line in `Task.toLi` on the main branch. The reason I changed this line to `document.createElement` is that `li.createSpan` is a shortcut in the Obsidian API, which the mock we use for running tests does not include, and I wanted the tests to correctly create the div and span structure. But it does exactly the same.

claremacrae commented on src/Task.ts:
```diff
@@ -438,159 +439,78 @@ export class Task {
-        }
-
-        return li;
+    /**
+     * Create an HTML rendered List Item element (LI) for the current task.
+     * @param {renderTails}
+     */
+    public async toLi(renderDetails: TaskLineRenderDetails): Promise<HTMLLIElement> {
+        return renderTaskLine(this, renderDetails);
```
There are some references to `Task.toLi` in CONTRIBUTING.

Could you please have a review of them and see if anything needs updating with the removal of code out to `renderTaskLine()`?

esm7 replied:
Will do.

claremacrae commented on src/TaskLineRenderer.ts:
```diff
@@ -0,0 +1,228 @@
+    li.classList.add('task-list-item', 'plugin-tasks-list-item');
+
+    const textSpan = document.createElement('span');
+    li.appendChild(textSpan);
+    textSpan.classList.add('tasks-list-text');
+    await taskToHtml(task, renderDetails, textSpan, textRenderer);
+
+    const checkbox = document.createElement('input');
+    li.appendChild(checkbox);
+    checkbox.classList.add('task-list-item-checkbox');
+    checkbox.type = 'checkbox';
+    if (task.status !== taskModule.Status.TODO) {
+        checkbox.checked = true;
+        li.classList.add('is-checked');
+    }
```
A small point, but 44 to 56 are a bit dense to read. A comment or two might help clarify the structure..

I think a part of it may be the pattern of moving from lines like this:

```ts
const checkbox = li.createEl('input');
```

to lines like this:

```ts
    const checkbox = document.createElement('input');
    li.appendChild(checkbox);
```

For my own education, what's the benefit of the new version?

esm7 replied:
Will add a few comments to explain the flow.
Regarding the change in how the various elements are created, `li.createEl` is a shortcut provided for convenience by the Obsidian API, but we don't have it available when running tests, and I wanted the tests to be able to create the full div and span structure, so had to convert all of these to the equivalent but more elaborate `document.createElement` and `appendChild` calls.

esm7 replied:
Actually there's more to that which is worth sharing. I'm dreaming of creating an Android home screen query widget that can do stuff like presenting today's tasks on the phone home screen etc, so in addition of solving a practical problem, I thought it's worthwhile to decrease the dependency of the rendering code on Obsidian's API.
Given the limitations of the free time I'm able to find this will likely stay in the dream scope, but making it a step more reachable in case someone else wants to do it looked worthwhile :)

claremacrae replied:
Thanks for both those helpful answers!

claremacrae replied:
Thanks again. I've added a comment in the code explaining why the more verbose form is used.

claremacrae commented on src/TaskLineRenderer.ts:
```diff
@@ -0,0 +1,228 @@
+        checkbox.checked = true;
+        li.classList.add('is-checked');
+    }
+    checkbox.addEventListener('click', (event: MouseEvent) => {
+        event.preventDefault();
```
Another interesting change that I totally trust but would like to learn from.

The previous code was:

```ts
checkbox.onClickEvent((event: MouseEvent) => {
```

The new code is consistent with that in `LivePreviewExtension`

```ts
        this.handleClickEvent = this.handleClickEvent.bind(this);
        this.view.dom.addEventListener('click', this.handleClickEvent);
```

What's the benefit of the change?
Is there any change in behaviour? (I assume not)

esm7 replied:
Same as above, `checkbox.onClickEvent` doesn't work without the Obsidian API (see it all under `global.HTMLElement` in `obsidian.d.ts`).

claremacrae commented on tests/TaskLineRenderer.test.ts:
```diff
@@ -0,0 +1,156 @@
+            const fullLayoutOptions = { ...new LayoutOptions(), ...layoutOptions };
+            const parentRender = await createMockParentAndRender(task, fullLayoutOptions);
+            const renderedDescription = getDescriptionText(parentRender);
+            expect(renderedDescription).toEqual(expectedRender);
```
When this `expect()` fails, there is no useful information that I can see about the location of the failure - it could be any one of a number of tests in the same block:

```text
at /Users/clare/Documents/develop/Obsidian/schemar/obsidian-tasks/tests/TaskLineRenderer.test.ts:101:41
at Generator.next (<anonymous>)
at fulfilled (/Users/clare/Documents/develop/Obsidian/schemar/obsidian-tasks/node_modules/tslib/tslib.js:115:62)
at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

See below for some other ideas that may help with this.

claremacrae commented on tests/TaskLineRenderer.test.ts:
```diff
@@ -0,0 +1,156 @@
+            expect(renderedDescription).toEqual(expectedRender);
+        };
+
+        const taskLine = '- [ ] Wobble ⏫ 📅 2022-07-02 ⏳ 2022-07-02 🛫 2022-07-02 🔁 every day';
```
Tests can be made more robust and useful by removing duplication in the input data.
For example, if someone broke the rendering of due date and made it write scheduled date instead, these tests would still pass.

I suggest using unique values for each of the dates in the initial task.

esm7 replied:
Will do.

claremacrae commented on tests/TaskLineRenderer.test.ts:
```diff
@@ -0,0 +1,156 @@
+
+        // Without start date
+        await testLayoutOptions(
+            taskLine,
+            { hideStartDate: true },
+            'Wobble ⏫ 🔁 every day ⏳ 2022-07-02 📅 2022-07-02',
+        );
+
+        // Without scheduled date
+        await testLayoutOptions(
+            taskLine,
+            { hideScheduledDate: true },
+            'Wobble ⏫ 🔁 every day 🛫 2022-07-02 📅 2022-07-02',
+        );
+
+        // Without due date
+        await testLayoutOptions(taskLine, { hideDueDate: true }, 'Wobble ⏫ 🔁 every day 🛫 2022-07-02 ⏳ 2022-07-02');
+
+        // Without recurrence rule
+        await testLayoutOptions(
+            taskLine,
+            { hideRecurrenceRule: true },
+            'Wobble ⏫ 🛫 2022-07-02 ⏳ 2022-07-02 📅 2022-07-02',
+        );
+
+        const doneTask = '- [x] Wobble ✅ 2022-07-02 ⏫ 📅 2022-07-02 ⏳ 2022-07-02 🛫 2022-07-02 🔁 every day';
+
+        // Done task - default layout
+        await testLayoutOptions(
+            doneTask,
+            {},
+            'Wobble ⏫ 🔁 every day 🛫 2022-07-02 ⏳ 2022-07-02 📅 2022-07-02 ✅ 2022-07-02',
+        );
+
+        // Done task - without done date
+        await testLayoutOptions(
+            doneTask,
+            { hideDoneDate: true },
+            'Wobble ⏫ 🔁 every day 🛫 2022-07-02 ⏳ 2022-07-02 📅 2022-07-02',
+        );
```
The comments here are useful in conveying the purpose of the test. I wondered if it would be possible to rework the code so that they could be moved to test names instead.

### Sharing `testLayoutOptions()` between `it` blocks

I found that if I moved `testLayoutOptions()` to just above this line...

```ts
    it('renders task components according to the given layout', async () => {
```

... then I could write individual test, each with a meaningful name.

For example I was then able to add this new test:

```ts
    it('should write placeholder message if a date is invalid', async () => {
        const taskLine = '- [ ] Task with invalid due date 📅 2023-13-02';
        await testLayoutOptions(taskLine, {}, 'Task with invalid due date 📅 Invalid date');
    });
```

I liked that a lot. I have been concerned for some time that there is no useful information in the `📅 Invalid date` text - but had no way of testing it.

### Easy TDD

I can imagine in future TDDing this to write a failing test like this, to show that I expect the invalid date to also be displayed:

```ts
    it('should write placeholder message with the date if a date is invalid', async () => {
        const taskLine = '- [ ] Task with invalid due date 📅 2023-13-02';
        await testLayoutOptions(taskLine, {}, 'Task with invalid due date 📅 Invalid date: 2023-13-02');
    });
```

### Alternative layout

I experimented with a different layout for the above tests - by inlining taskLine:

```ts
    it('should write placeholder message if a date is invalid', async () => {
        await testLayoutOptions(
            '- [ ] Task with invalid due date 📅 2023-13-02',
            {},
            'Task with invalid due date 📅 Invalid date',
        );
    });
```

The Task text being a bit closer to vertically aligned means I find it easier to quickly glance and see how the input differs from the output.

Another advantage of inlining the input data for the test is that the input is right next to the expected value.

In the current tests, the `taskLine` value is used in many tests, which means that:

- It has to contain all the input data for all the places it might be used, meaning the test purpose is less clear than if specifically useful data were chosen for each little test.
- If there is a test failure, the input is a long way from the failure
- It encourages writing long `it` blocks with a single description, which makes it harder to locate any failure in the source.

claremacrae replied:
Looking again at this code from my comment above... as well test name adding info, it helps when the task description itself can say more about the purpose.

```ts
    it('should write placeholder message if a date is invalid', async () => {
        const taskLine = '- [ ] Task with invalid due date 📅 2023-13-02';
        await testLayoutOptions(taskLine, {}, 'Task with invalid due date 📅 Invalid date');
    });
```

esm7 replied:
Will do and restructure; I'm unsure if this is something you already did in your branch, so waiting to see the result of the merge.

claremacrae commented:
Hi @esm7 I was so excited about this I couldn't wait and have gone ahead and reviewed it and tested it! 🤣 

Here's a bit of a summary of what I did..

## Merged main and pushed, to make a build

- Created a branch with merged code: [GitHub - obsidian-tasks-group/obsidian-tasks at refactor-task-rendering-clare-rendering](https://github.com/obsidian-tasks-group/obsidian-tasks/tree/refactor-task-rendering-clare-rendering)
    - Build good: [Merge branch 'main' into refactor-task-rendering-clare-rendering · obsidian-tasks-group/obsidian-tasks@3effd76 · GitHub](https://github.com/obsidian-tasks-group/obsidian-tasks/actions/runs/3731324150)
    - A zip of the Tasks-Demo vault, with the new code (including main merged in) can be downloaded from the above GitHub Actions page.

## About merging from main.

I'm happy to be the person that merges main in to your branch, before merging your branch to main.

It took me 2 or 3 tries to get a working code from the merge, and it was only this quick because I understand all the changes made on main since the version you were starting from.

## Manual testing

I set up two vaults side-by-side. 

- This build - of main with - in the left:
    - [refactor: TagsField now implements 'sort by tag' (#1411) · obsidian-tasks-group/obsidian-tasks@61c3958 · GitHub](https://github.com/obsidian-tasks-group/obsidian-tasks/actions/runs/3730513870)
- And this build with the PR code and the merge from main on the right:
    - [Merge branch 'main' into refactor-task-rendering-clare-rendering · obsidian-tasks-group/obsidian-tasks@3effd76 · GitHub](https://github.com/obsidian-tasks-group/obsidian-tasks/actions/runs/3731324150)

Here's what I did with those two vaults - pretty much all using existing pages in the Tasks-Demo vault.

- Confirmed `explain` still works (as it was done on main after this PR branch was created)
- Checked both the backlink styling snippets in sample vault worked
- Checked that SlrVb's Alternative Checkboxes still worked
- Saved as PDF and diffed the two outputs
    - Wrote a query to show every task the test vault, with urgency enabled
    - In both vaults, I exported the Reading view as PDF
    - Then opened both PDFs in one Acrobat window
    - And went down one page at a time, switching between the two files
    - Any slight differences in the rendering would have stood out.
    - There were no differences!
- Tried a bunch of different show/hide options - see below.
- Did some inspecting of HTML in the developer tools and couldn't find any differences

For exploratory testing the new code, I incrementally worked through adding the following lines, one by one - and comparing left and right:

````text
```tasks
explain

show urgency

hide backlink
hide scheduled date
hide start date
hide due date
hide done date
hide priority

hide edit button
hide recurrence rule

show start date
```
````

claremacrae commented:
Note to self:

Ooooh.. I was going to say that my branch mentioned above isn't suitable for merging, as it contains some cruft from the review process. Like the only way I could figure out how to diff the code removed from Task with the code adding in a new file was by using a temporary intermediate file, which I rearranged until the order matched that in the new code, so I could diff it.

But... because this project squashes all commits when merging, if I remove all traces of my edits from my branch and then merge that, the end result will be fine to merge.

I could then clean up the commit messages slightly at the point when I merge to main.

esm7 commented:
Just wanted to drop a note here saying that I'm having a super busy day, but I was excited to skim through the comments, and either tomorrow or on Wednesday I'll reply to everything!

esm7 commented:
> > ## How has this been tested?
> > Tests were added to the rendering flow.
> 
> Could you give me an idea of how much testing you've done to confirm that the appearance is indeed unchanged please? For example, what kinds of combinations of layout options have you tested with? Thanks.

Maybe this question is no longer relevant since you discussed very thorough tests later on, but I did a comparison similar to what you described: tested all the layout options I could think of and compared the resulting DIV structure and look.

claremacrae commented:
If not too late... @esm7 I wonder if it's worth your merging my branch in to yours before you respond to any feedback?

esm7 commented:
> I'm not sure that `src/Query/` is the best location for `TaskLayout` and `TaskLineRenderer`
> 
>     * I wondered about creating `src/Query/Layout/`
> 
>     * That name is consistent with `layout` [in the docs](https://obsidian-tasks-group.github.io/obsidian-tasks/queries/layout/).
> 
>     * **But** I am conscious that you have other PRs coming, so we should leave them where they are until all your changes are merged.

Will move it to wherever you decide eventually, just note that:
1. I am intentionally sparse when it comes to directory structure and nesting. From my experience an overly-deep nesting of source files can make it difficult to navigate a code base, and in my own company and projects that I manage I usually instruct developers to use sub-directories mostly only where encapsulation is needed. In Tasks I think that a good example of that is the `Query/Filter` directory: it hides a big amount of details that aren't needed for understanding the higher-level `Query` mechanism. (Either way: this was just a side note, and eventually will stick to the conventions of this project, of course!)
2. Regarding the next PR that is coming and its relation to the above, you might be surprised how little is needed on top of this refactor to support extensive custom layouts. I'd say 80% of the work is in this PR, and the next one just presents changes to the `div`/`span` structure of the rendering and adds custom classes, which is really simple with the new code structure.

esm7 commented:
> If not too late... @esm7 I wonder if it's worth your merging my branch in to yours before you respond to any feedback?

This is exactly the right moment! I was trying to figure out how to get your changes exactly when I saw this reply :)

esm7 commented:
> 

I'm incredibly happy that you were the one to merge from main. You saved me a big headache, and like you say, being familiar with all the changes makes you the best person for the job.

esm7 commented:
@claremacrae just making sure you're planning to merge your changes to my branch soon, and update here when you do so I can continue :pray:

claremacrae commented:
Oh I'm glad you asked.

> If not too late... @esm7 I wonder if it's worth your merging my branch in to yours before you respond to any feedback?

I was presuming you would be able to merge them yourself - and didn't know you were waiting for me.

I'll see if I can create a pull request - or maybe push it directly myself.

claremacrae commented:
> @claremacrae just making sure you're planning to merge your changes to my branch soon, and update here when you do so I can continue 🙏

@esm7 It appears I have merged in to this PR successfully... sorry for all the spam emails along the way!

esm7 commented:
@claremacrae I just pushed an update which addresses all your comments, I believe.

claremacrae commented:
> @claremacrae I just pushed an update which addresses all your comments, I believe.

Thank you! From an initial look on my phone it looks good.  I’ll work through and resolve the comments later.

Do you remember why the change to yarn.lock in this commit? Am asking as it has introduced a new merge conflict.

esm7 commented:
> > @claremacrae I just pushed an update which addresses all your comments, I believe.
> 
> Thank you! From an initial look on my phone it looks good. I’ll work through and resolve the comments later.
> 
> Do you remember why the change to yarn.lock in this commit? Am asking as it has introduced a new merge conflict.

Oh, I remember noting to myself not to commit that and turns out I did eventually...

claremacrae commented:
OK thanks, no worries, I'll figure out a way to revert it before I merge.

claremacrae commented:
Hi @esm7 All now merged - great work, thank you very much indeed!

esm7 commented:
Great to see this merged. Next step is to make the full "themeability" branch build on top of this, which is the easy part, and then document it thoroughly, which is a little harder :)
Hoping to have time for that next week.
